[FEAT] 투두추가화면과 관련된 UI 프로토타입 구현

dlrjswns opened the pull request:
## 📌 PR 요약

🌱 작업한 내용
- 피그만에 존재하는 투두추가화면에 관련된 UI를 구현하였습니다.
- 체크여부에 따라서 취소선 여부를 추가하도록 구현하였습니다.
- 캘린더를 통해 선택한 날짜에 대한 값을 라벨에 보여지도록 하였습니다.

🌱 PR 포인트
- 이후 API연동은 따로 pr올리도록 하겠습니다 !!

## 📸 동작영상
https://user-images.githubusercontent.com/39263235/237008297-57b56062-5a18-4582-850d-23e6935c0d6b.mp4


## 📮 관련 이슈
- Resolved: #355

dlrjswns committed:
[ADD] 투두화면에 사용될 캘린터라이브러리 추가

dlrjswns committed:
[ADD] 투두추가화면을 위한 클래스 UI 추가

dlrjswns committed:
[ADD] 투두추가화면에 대한 뷰 컨트롤러 추가

dlrjswns committed:
[ADD] 텍스트필드에 밑줄선을 긋는 공통함수 추가

dlrjswns committed:
[FEAT] 투두텍스트필드에 밑줄선을 추가하는 코드

dlrjswns committed:
[FEAT] 추가하고싶은 투두의 양이 많을 경우 스크롤가능케하기위해 스크롤 뷰 추가

dlrjswns committed:
[FEAT] 투두추가를 위한 UI 및 모델 정의

dlrjswns committed:
[ADD] 투두추가화면에 사용될 이미지 에셋 추가

dlrjswns committed:
[ADD] 투두뷰 화면 구성을 위한 UI 추가

dlrjswns committed:
[FEAT] 투두추가화면 캘린터 UI 추가

dlrjswns committed:
[FEAT] 오늘을 나타내는 날짜를 stroke형식으로 표현

dlrjswns committed:
[FEAT] 선택한 날짜에 대한 텍스트를 받기위한 코드 추가

dlrjswns committed:
[ADD] 달력에서 선택한 날짜를 전달하기위한 코드 추가

dlrjswns committed:
[ADD] 텍스트라벨에 속성 추가

dlrjswns committed:
[CHORE] 해당 날짜에 관련된 투두가 존재하지않을 경우 디폴트값 지정을 위한 코드 수정

WhiteHyun commented on PLUB/Sources/Views/Home/MainPage/AddTodoViewController.swift:
```diff
@@ -7,11 +7,145 @@
+      $0.height.equalTo(29)
+    }
+    
+    calendarView.snp.makeConstraints {
+      $0.height.equalTo(357.29)
+    }
+    
+    containerStackView.setCustomSpacing(13.71, after: calendarView)
+  }
```
Figma에서 소수점이 나오는 거는 대부분 디자이너 실수여서, 직접 여쭤보시거나 반올림하셔서 정수로 설정하시면 되세요!

dlrjswns replied:
아 그런가요 ?? 357로 수정해서 올리겠습니다 감사합니다 👍👍

WhiteHyun commented on PLUB/Sources/Views/Home/MainPage/Component/AddTodoView.swift:
```diff
@@ -0,0 +1,195 @@
+    
+    override func layoutSubviews() {
+      super.layoutSubviews()
+      if type == .input {
+        todoTextField.addUnderline(color: .lightGray)
+      }
```
layoutSubviews는 여러번 호출될 수 있는 메서드라서 todoTextField에 의도치 않게 많은 underline layer가 추가될 수도 있어보여요(즉 리소스 낭비). 이 부분을 고민해보면 좋을 것 같습니다.

dlrjswns replied:
아 저 부분은 프로퍼티속성이나 내부 함수에서 구현하면은 적용이 안되서 해당 함수안에 넣었는데 한번 다시 확인해볼게요 잠시만요 !!

dlrjswns replied:
일단 확인하고왔는데 제가 생각할 수 있는 방법에선 layoutSubviews에서만 되는거같긴한데 혹시 괜찮은 아이디어가 있으실가요 ??

WhiteHyun replied:
separatorLine이라는 UIView를 추가해서 constraint작업하면 어떨까요?

dlrjswns replied:
> layoutSubviews는 여러번 호출될 수 있는 메서드라서 todoTextField에 의도치 않게 많은 underline layer가 추가될 수도 있어보여요(즉 리소스 낭비). 이 부분을 고민해보면 좋을 것 같습니다.

일단 취소선이 layer를 건드는 작업이라 shouldRasterize라는 속성을 추가하였습니다.
해당 속성은 처음만 layer작업을 렌더링하고 이후 작업은 재사용하여 다음과 같은 경우에 성능향상을 기대할 수 있을거같다는 판단에 추가해보았습니다

ps. 사실 그냥 bottomView하나 만들어서 하게되면 layoutSubviews를 쓸 필요가 없을꺼같긴한데 수빈님하고 승현님은 어떻게 생각하실가요 ?? @WhiteHyun 홍 @soobin-k 김

WhiteHyun replied:
취소선이아니라 밑줄선 아닌가용..?

dlrjswns replied:
> 취소선이아니라 밑줄선 아닌가용..?

아아 정정하겠습니다 밑줄선입니다 ㅋㅋㅋㅋㅋㅋ

dlrjswns replied:
> separatorLine이라는 UIView를 추가해서 constraint작업하면 어떨까요?

아 그렇다면 하나 뷰를 만들어서 작업하도록 하겠습니다 !! UIView보다 CALayer를 추가하는 작업이 더 가볍다고 하더라구요 !! 그래서 기존 방법을 고수하되 방법도 한번 찾아봐야겠습니다 🧐🧐

WhiteHyun replied:
> 일단 취소선이 layer를 건드는 작업이라 shouldRasterize라는 속성을 추가하였습니다. 해당 속성은 처음만 layer작업을 렌더링하고 이후 작업은 재사용하여 다음과 같은 경우에 성능향상을 기대할 수 있을거같다는 판단에 추가해보았습니다

만약 under line layer가 존재하는 상태에서 shouldRasterize를 `true`를 적용하면 성능 효과를 발휘할 수 있을 것 같아요. 근데, 현재 코드에서는 계속해서 새로운 layer를 추가하는 거라, shouldRasterize 속성을 바꿔도 기존과 동일하게 동작할 것 같아 보입니다..!

저도 이 속성에 대해 잘 알지 못해 제 답이 정답은 아닐 수 있습니다.

soobin-k replied:
저도 UIView를 추가하는 방법이 제일 좋을 것 같습니다!
디자인 가이드상으로 따로 bottomLineView가 유동적으로 사라지거나 변형되지 않고 최초에만 세팅해주면 된다고 생각했어요.

dlrjswns committed:
[CHORE]  pr리뷰에 따른 값 수정

dlrjswns committed:
[ADD] 텍스트필드에 대한 속성을 추가하였습니다.

layer.shouldRasterize라는 속성은 처음 한 번만 렌더링 하고 이후 애니메이션부턴 재활용되기에 화면내에서 움직이기는 하나 모양이 변하지않을 때 성능향상을 기대할 수 있는 속성이라합니다.

dlrjswns committed:
[CHORE] 텍스트필드 바텀라인추가에 대한 코드 수정

dlrjswns committed:
[DEL] Ex_UITextField 코드 삭제

dlrjswns committed:
[DEL] 테스트코드 삭제

dlrjswns committed:
[CHORE] 프로퍼티 코드 수정

WhiteHyun commented on PLUB/Sources/Views/Home/MainPage/Component/AddTodoView.swift:
```diff
@@ -0,0 +1,196 @@
+    configureUI()
+    completionHandler = { [weak self] date in
+      let dateString = DateFormatterFactory.todolistDate.string(from: date)
+      if Calendar.current.isDateInToday(date) {
+        self?.dateLabel.text = "\(dateString) (오늘)"
+      } else {
+        self?.dateLabel.text = dateString
+      }
```
디자인 상에서는 금일이 아닌 경우 글자 색상이 검은색으로 나와있는데, 확인 부탁드립니다.

|Figma design|건준님이 구현해주신 디자인|
|:-:|:-:|
|![image](https://github.com/PLUB2022/PLUB-iOS/assets/57972338/c5e1ec75-833a-4813-a70d-18874f416a5c)|<img width="566" alt="image" src="https://github.com/PLUB2022/PLUB-iOS/assets/57972338/3c9b450a-9013-47f9-aae5-bf4e26af832e">|

soobin-k commented on PLUB/Sources/Views/Home/MainPage/Component/AddTodoView.swift:
```diff
@@ -0,0 +1,196 @@
+    private let disposeBag = DisposeBag()
+    private let type: TodoViewType
+    
+    private lazy var emptyCheckView = UIView().then {
```
lazy var를 사용하신 이유가 궁금합니다!

WhiteHyun replied:
이거는 제가 답해드릴 수 있을 것 같습니다.
AddTodoView에서는 TodoViewType에 따라 보이는 뷰가 다릅니다.
그래서 타입별로 모든 View를 초기화 하지 않고, 나중에 타입에 맞춰서 지연 초기화를 적용하여 메모리상 이점을 얻으려고 하는 것 같습니다.
실제로 135줄부터 보시면 타입에 따라 뷰와 제약조건 설정을 하는 것을 볼 수 있습니다.

soobin-k replied:
아 type 별로 ui가 달라지는 것 때문에 저렇게 지연 초기화를 사용한거군요..! 이해했습니다!! 답변 감사드려요 승현님!!👍

dlrjswns replied:
승현님 역시 마음이 통했네요 ❤️

soobin-k commented on PLUB/Sources/Views/Home/MainPage/Component/AddTodoView.swift:
```diff
@@ -0,0 +1,196 @@
+      contentLabel.attributedText = model.isChecked ? contentLabel.text?.strikeThrough() : NSAttributedString(string: contentLabel.text ?? "")
+    }
+    
+    func bind() {
```
private 붙여주면 좋을 것 같아요!

dlrjswns replied:
헉 놓친 부분이 있었네요 수정하겠습니다 !!

soobin-k reviewed (approved):
고생많으셨습니다 건준님!👍
다음 pr에 키보드 관련 처리도 작업해주시면 더 좋을 것 같아요😊

dlrjswns commented:
> 고생많으셨습니다 건준님!👍 다음 pr에 키보드 관련 처리도 작업해주시면 더 좋을 것 같아요😊

아 맞네요 그 부분도 인지하고 있겠습니다 😮😮

dlrjswns committed:
[CHORE] private 접근제어자 추가
